### PR TITLE
Ensure h2 streams are fully received before closing (#1245)

### LINKED ISSUE #1245: H2 connection starts to deadline every request on a connection after random interval
We are mirroring some gRPC production traffic through linkerd and after a random interval (anywhere from 30 seconds to 2 hours) every request over the connection starts to deadline. After this point all of the request charts in the admin UI go to zero. If the connection is recreated through either restarting linkerd or restarting the client service, traffic flow is temporarily restored. 

There are no abnormal log messages with `TRACE` verbosity turned on when linkerd gets in this state. 

- We did a tcpdump to verify that the traffic was reaching linkerd (it is). We observe that the client service is sending h2 request frames (`DATA`, `HEADERS`) over the stream, and then after the deadline interval sends a `RST_STREAM`, which is intended behavior. Nothing during this interval is sent from linkerd other than TCP ACK packets. 
- We also ran a test without linkerd to see if it works (it does).
- We tried disabling failure accrual on the client linkerd and the server linkerd, it didn't make a difference.

Metrics snapshot below.

### PR DESCRIPTION
**Pull request by klingerf**
## Problem

A race condition existed that would cause a stream to be closed by the remote sender before the local receiver had fully received all frames in an h2 stream. This would result in the final frame of an h2 response never being sent back to the caller, causing the caller's connection to hang indefinitely.

## Solution

Make the receiver fully responsible for closing streams, to ensure that they are fully received before closing.

## Validation

I've deployed this branch to a test environment and it has successfully served gRPC traffic for the past 48 hours, whereas without this fix the connection hangs within about 20 minutes.

As part of this change I've included some small refactors for readability in the Netty4StreamTransport. Will also try to add a unit test in a follow-up commit. Fixes #1245.

### CODE REVIEW

**Comment by mattolson** [comment]
🙏

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala [hunk 1]
```diff
@@ -187,28 +184,28 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
...
       case f@Failure(_) if f.isFlagged(Failure.Rejected) =>
-        log.debug(f, "[%s] remote message interrupted", prefix)
+        log.debug("[%s] remote message interrupted %s", prefix, f)
         localReset(Reset.Refused)
 
       case e =>
-        log.debug(e, "[%s] remote message interrupted", prefix)
+        log.debug("[%s] remote message interrupted %s", prefix, e)
         localReset(Reset.InternalError)
     }
```
Comment (olix0r) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: the prior format is used so that we print stack traces when they are enabled (this can be controlled at start time via a system property).

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala [hunk 2]
```diff
@@ -551,7 +568,7 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
     p
   }
 
-  private[this] val writeFrame: Frame => Future[Unit] = { frame =>
+  private[this] def writeFrame(frame: Frame): Future[Unit] =
```
Comment (olix0r) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: why is this necessary?  We prefer to use val'd functions when possible, since they save on unnecessary anonymous function allocations.

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala [hunk 3]
```diff
@@ -488,7 +505,7 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
     val streamFF = headersF.map(_ => writeStream(msg.stream))
 
     val writeF = streamFF.flatten
-    onReset.onFailure(writeF.raise(_))
+    onReset.onFailure(writeF.raise)
```
Comment (olix0r) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: the prior form is used to be explicit that this allocates.  We use val'd functions and the paren-less form to indicate that there is no allocation performed, and we use parens to illustrate that an anonymous function is allocated (i.e. when we call a function rather than pass a val'd function).

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala [hunk 4]
```diff
@@ -435,29 +441,40 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
...
+            if (resetFromLocal(remote, Reset.InternalError)) false
+            else recv(data)
+
+          case Open(remote@RemoteStreaming()) =>
+            if (recvFrame(toFrame(data), remote)) true
+            else recv(data)
+
+          case LocalClosed(remote@RemoteStreaming()) =>
+            if (recvFrame(toFrame(data), remote)) true
+            else recv(data)
```
Comment (olix0r) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: Is this change related to the bugfix in some way or is this just a refactor?

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala [hunk 5]
```diff
@@ -274,7 +271,6 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
 
       case state@RemoteClosed() =>
         if (stateRef.compareAndSet(state, Closed(Reset.NoError))) {
-          state.close()
```
Comment (olix0r) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: Is it worth adding a comment here explaining how state.close is expected to be called?

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala (hunk 1 above)
Comment (klingerf) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: Ah, this is good to know. Which system property enables stack traces? Does that also world for Failure types (cases 3 and 4)? I can revert this change.

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala (hunk 4 above)
Comment (klingerf) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: This is just a refactor. I kept getting confused that we process `Http2HeadersFrame.isEndStream` and `Http2HeadersFrame` in separate cases, but we process all `Http2DataFrame` in the same case. Breaking data frames out into two separate cases feels more consistent with how we're handling header frames.

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala (hunk 3 above)
Comment (klingerf) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: Ok, good to know. Will revert.

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala (hunk 2 above)
Comment (klingerf) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: I changed this for consistency with the `writeHeaders` and `writeStream` defs defined above. I can revert this, but I might also convert those other defs to vals.

File: finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala [hunk 6]
```diff
@@ -442,8 +442,10 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
 
           case Open(remote@RemoteStreaming()) =>
             if (stateRef.compareAndSet(state, remote.toRemoteClosed)) {
-              if (recvFrame(toFrame(data), remote)) true
-              else throw new IllegalStateException("stream queue closed prematurely")
+              if (recvFrame(toFrame(data), remote)) {
+                remote.close()
+                true
+              } else throw new IllegalStateException("stream queue closed prematurely")
```
Comment (zackangelo) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: Should there be `remote.close()` here as well if the stream queue is closed?

Comment (klingerf) [inline finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala]: If `recvFrame` returns false, that means that the call to `remote.offer` failed, due to the stream already being closed. So I think in this case calling `remote.close` again would be redundant.
